docs/_includes/Expressions.md

```diff
@@ -1121,4 +1121,37 @@ Expressions =
 The ``Expressions`` non-terminal represents a list of
 one or more expressions separated by a comma.
 
+## Compile-Time Constants {#sec-compile-time-constants}
 
+In certain situations in Dafny it is helpful to know what the value of a 
+constant is during program analysis, before verification or execution takes
+place. For example, a compiler can choose an optimized representation of a 
+`newtype` that is a subset of `int` if it knows the range of possible values 
+of the subset type: if the range is within 0 to 255 (inclusive), then an 
+unsigned 8-bit representation can be used.
+
+To continue this example, suppose a new type is defined as
+```
+const MAX := 47
+newtype mytype = x | 0 <= x < MAX*4
+```
+In this case, we would prefer that Dafny recognize that `MAX*4` is 
+known to be constant with a value of `188`. The kinds of expressions 
+for which such an optimization is possible are called
+_compile-time constants_. Note that the representation of `mytype` makes
+no difference at all to verification, only to the runtime performance of
+compiled code. In addition, though, using a symbolic constant (which may 
+well be used elsewhere as well) improves the self-documentation of the code.
+
+In Dafny, the following expressions are compile-time constants[^CTC], recursively 
+(that is, the arguments of any operation must themselves be compile-time constants):
+- int, real, boolean, and string literals
+- builtin unary operations: `-` (int and real), `!` (bool and int), and `|...|` (string length)
+- binary operations: `+ - * / %` (int and real), `<< >>` (int), `+` (string concatenation), `&& ||` (bool)
+- bit operations: `& | ^` (int)
+- comparison operations: `< <= > >=` (int and real), `== !=` (int, real, bool, string)
```

Comment:
String equality/disequality was not handled above. (For `.ResolvedOp`, they are called `SeqEq` and `SeqNeq`. For `.Op`, they are `Eq` and `Neq`.) Note that if in C# you write `a == b` where `a` and `b` are expressions of static type `object`, then you'll get pointer equality even when `a` and `b` evaluate to values of type `string`. To get string equality, you must do `(string)a == (string)b`.